File: backend/Services/AuthService.cs
using backend.Common;
using backend.DTOs;
using backend.Interfaces;
using backend.Models;
using Microsoft.AspNetCore.Identity;
using Microsoft.EntityFrameworkCore;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Security.Cryptography;
using System.Text;

namespace backend.Services
{
    public class AuthService : IAuthService
    {
        private readonly UserManager<ApplicationUser> _userManager;
        private readonly SignInManager<ApplicationUser> _signInManager;
        private readonly IEmailService _emailService;
        private readonly IConfiguration _configuration;


        public AuthService(
            UserManager<ApplicationUser> userManager,
            IEmailService emailService,
            SignInManager<ApplicationUser> signInManager,
            IConfiguration configuration)
        {
            _userManager = userManager;
            _signInManager = signInManager;
            _emailService = emailService;
            _configuration = configuration;
        }

        //Register
        public async Task<AuthDTO.AuthResponseDTO> RegisterAsync(AuthDTO.RegisterDTO dto)
        {
            var existing = await _userManager.FindByEmailAsync(dto.Email.Trim().ToLower());
            if (existing != null)
            {
                //Soft-deleted accounts have their email anonymised to deleted_xxx@deleted.loanit
                //This guard catches any edge case where it does.
                if (existing.IsDeleted)
                    throw new ArgumentException("This email was previously used on a deleted account. Please contact support.");

                throw new ArgumentException("An account with this email already exists.");
            }

            //Check if username is already taken
            var usernameTaken = await _userManager.FindByNameAsync(dto.Username.Trim());
            if (usernameTaken != null)
                throw new ArgumentException("That username is already taken.");


            var user = new ApplicationUser
            {
                FullName = dto.FullName.Trim(),
                Email = dto.Email.Trim().ToLower(),
                UserName = dto.Username.Trim(),
                Address = dto.Address.Trim(),
                Latitude = dto.Latitude,
                Longitude = dto.Longitude,
                DateOfBirth = dto.DateOfBirth,
                Gender = dto.Gender?.Trim(),
                AvatarUrl = dto.AvatarUrl?.Trim(),
                Score = 100,
                MembershipDate = DateTime.UtcNow
            };

            var result = await _userManager.CreateAsync(user, dto.Password);
            if (!result.Succeeded)
            {
                var errors = string.Join(", ", result.Errors.Select(e => e.Description));
                throw new ArgumentException(errors);
            }

            await _userManager.AddToRoleAsync(user, Roles.User);

            //Send confirmation email — user cant log in until email is confirmed
            var confirmToken = await _userManager.GenerateEmailConfirmationTokenAsync(user);
            var confirmLink = $"{_configuration["App:BaseUrl"]}/api/auth/confirm-email?userId={user.Id}&token={Uri.EscapeDataString(confirmToken)}";

            await _emailService.SendEmailAsync(
                user.Email,
                "Confirm your LoanIt account",
                $"<h2>Welcome to LoanIt, {user.FullName}!</h2>" +
                $"<p>Please confirm your email address by clicking the link below:</p>" +
                $"<p><a href='{confirmLink}'>Confirm Email</a></p>" +
                $"<p>If you did not create an account, you can ignore this email.</p>"
            );

            //Return a response without a token — ""Angular should show "check your email" screen""
            return new AuthDTO.AuthResponseDTO
            {
                UserId = user.Id,
                FullName = user.FullName,
                Username = user.UserName,
                Email = user.Email,
                Role = Roles.User
                //Token is intentionally empty — login is blocked until email is confirmed
            };
        }

        //Confirm email

        public async Task<bool> ConfirmEmailAsync(string userId, string token)
        {
            var user = await _userManager.FindByIdAsync(userId);
            if (user == null) return false;

            var result = await _userManager.ConfirmEmailAsync(user, token);
            return result.Succeeded;
        }

        //Login
        public async Task<AuthDTO.AuthResponseDTO> LoginAsync(AuthDTO.LoginDTO dto)
        {
            var user = await _userManager.FindByEmailAsync(dto.Email.Trim().ToLower());

            if (user == null)
                throw new ArgumentException("Invalid email or password.");

            if (user.IsDeleted)
                throw new ArgumentException("This account has been deleted.");

            if (!await _userManager.IsEmailConfirmedAsync(user))
                throw new ArgumentException("Please confirm your email before logging in.");

            var result = await _signInManager.CheckPasswordSignInAsync(user, dto.Password, lockoutOnFailure: true);

            if (result.IsLockedOut)
                throw new ArgumentException("Account is locked. Please try again in 10 minutes.");

            if (!result.Succeeded)
                throw new ArgumentException("Invalid email or password.");

            return await BuildAuthResponseAsync(user);
        }

        //Refresh token
        public async Task<AuthDTO.AuthResponseDTO> RefreshTokenAsync(AuthDTO.RefreshTokenDTO dto)
        {
            //Hash the incoming token and look up by hash — never store plain text
            var tokenHash = HashToken(dto.RefreshToken);

            var user = await _userManager.Users
                .FirstOrDefaultAsync(u => u.RefreshToken == tokenHash);

            if (user == null)
                throw new UnauthorizedAccessException("Invalid refresh token.");

            if (user.IsDeleted)
                throw new UnauthorizedAccessException("This account has been deleted.");

            if (user.RefreshTokenExpiry == null || user.RefreshTokenExpiry < DateTime.UtcNow)
                throw new UnauthorizedAccessException("Refresh token has expired. Please log in again.");

            return await BuildAuthResponseAsync(user);
        }

        //Logout
        public async Task LogoutAsync(string userId)
        {
            var user = await _userManager.FindByIdAsync(userId);
            if (user == null) return;

            user.RefreshToken = null;
            user.RefreshTokenExpiry = null;

            await _userManager.UpdateAsync(user);
        }

        //Forgot password
        public async Task<bool> ForgotPasswordAsync(string email)
        {
            var user = await _userManager.FindByEmailAsync(email.Trim().ToLower());

            //Always return true — never reveal whether an email exists in the system
            if (user == null) return true;

            var token = await _userManager.GeneratePasswordResetTokenAsync(user);
            var resetLink = $"{_configuration["App:BaseUrl"]}/reset-password?email={Uri.EscapeDataString(email)}&token={Uri.EscapeDataString(token)}";

            await _emailService.SendEmailAsync(
                email,
                "Reset your LoanIt password",
                "<h2>Password Reset</h2>" +
                "<p>We received a request to reset the password for your LoanIt account.</p>" +
                $"<p><a href='{resetLink}'>Reset Password</a></p>" +
                "<p>This link expires in 1 hour. If you did not request a password reset, you can ignore this email.</p>"
            );

            return true;
        }

        //Reset password
        public async Task<bool> ResetPasswordAsync(AuthDTO.ResetPasswordDTO dto)
        {
            var user = await _userManager.FindByEmailAsync(dto.Email.Trim().ToLower());
            if (user == null) return false;

            var result = await _userManager.ResetPasswordAsync(user, dto.Token, dto.NewPassword);
            return result.Succeeded;
        
        }

        //Resend confirmation email
        public async Task ResendConfirmationEmailAsync(string email)
        {
            var user = await _userManager.FindByEmailAsync(email.Trim().ToLower());

            //Always return silently — never reveal whether the email exists
            if (user == null) return;

            if (user.IsDeleted) return;

            if (await _userManager.IsEmailConfirmedAsync(user)) return;

            var confirmToken = await _userManager.GenerateEmailConfirmationTokenAsync(user);
            var confirmLink = $"{_configuration["App:BaseUrl"]}/api/auth/confirm-email?userId={user.Id}&token={Uri.EscapeDataString(confirmToken)}";

            await _emailService.SendEmailAsync(
                user.Email!,
                "Confirm your LoanIt account (resent)",
                $"<h2>Email Confirmation</h2>" +
                $"<p>Here is your new confirmation link for your LoanIt account:</p>" +
                $"<p><a href='{confirmLink}'>Confirm Email</a></p>" +
                $"<p>If you did not request this, you can ignore this email.</p>"
            );
        }



        //JWT + refresh token builder
        private async Task<AuthDTO.AuthResponseDTO> BuildAuthResponseAsync(ApplicationUser user)
        {
            var roles = await _userManager.GetRolesAsync(user);
            var role = roles.FirstOrDefault() ?? Roles.User;
            var jwt = _configuration.GetSection("Jwt");
            var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(jwt["Key"]!));
            var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
            var expiresAt = DateTime.UtcNow.AddHours(1);

            var claims = new[]
            {
                new Claim(JwtRegisteredClaimNames.Sub, user.Id),
                new Claim(JwtRegisteredClaimNames.Email, user.Email!),
                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString()),
                new Claim(ClaimTypes.NameIdentifier, user.Id),
                new Claim(ClaimTypes.Name, user.FullName),
                new Claim(ClaimTypes.Role, role),
                new Claim("score", user.Score.ToString()),
                new Claim("isVerified", user.IsVerified.ToString().ToLower()),
                new Claim("securityStamp", user.SecurityStamp ?? string.Empty) //Security stamp — if user changes password, all existing JWTs are immediately invalid


            };

            var token = new JwtSecurityToken(
                issuer: jwt["Issuer"],
                audience: jwt["Audience"],
                claims: claims,
                expires: expiresAt,
                signingCredentials: creds
            );

            //Rotate refresh token on every use — old token immediately invalidated
            var rawRefreshToken = GenerateRefreshToken();
            user.RefreshToken = HashToken(rawRefreshToken);
            user.RefreshTokenExpiry = DateTime.UtcNow.AddDays(7);

            //if this update fails, throw before returning the token
            //so we never hand out a token the DB doesn't recognise
            var updateResult = await _userManager.UpdateAsync(user);
            if (!updateResult.Succeeded)
                throw new InvalidOperationException("Failed to persist refresh token. Please try again.");

            return new AuthDTO.AuthResponseDTO
            {
                Token = new JwtSecurityTokenHandler().WriteToken(token),
                RefreshToken = rawRefreshToken,
                UserId = user.Id,
                FullName = user.FullName,
                Username = user.UserName!,
                Email = user.Email!,
                Role = role,
                Score = user.Score,
                UnpaidFinesTotal = user.UnpaidFinesTotal,
                ExpiresAt = expiresAt
            };
        }

        //Generates a cryptographically random 64-byte base64 string
        private static string GenerateRefreshToken()
        {
            var bytes = new byte[64];
            using var rng = RandomNumberGenerator.Create();
            rng.GetBytes(bytes);
            return Convert.ToBase64String(bytes);
        }

        public static string HashToken(string token)
        {
            var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(token));
            return Convert.ToBase64String(bytes);
        }


    }


}

File: frontend/src/app/services/auth-service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { AuthDTO } from '../dtos/authDTO';
import { ApiResponseDTO } from '../dtos/apiResponseDTO';

@Injectable({
  providedIn: 'root',
})
export class AuthService {
  private readonly baseUrl = 'https://localhost:7183/api/auth';

    private _scheduleWarning?: () => void;
    setScheduler(fn: () => void): void { this._scheduleWarning = fn; }
 

  constructor(private http: HttpClient) {}

  //POST /api/auth/register
  register(dto: AuthDTO.RegisterDTO): Observable<AuthDTO.AuthResponseDTO> {
    return this.http
      .post<AuthDTO.AuthResponseDTO>(`${this.baseUrl}/register`, dto)
      .pipe(tap((res) => { this.saveTokens(res.token, res.refreshToken); this._scheduleWarning?.(); }));
  }

  //GET /api/auth/confirm-email?userId=&token=
  confirmEmail(userId: string, token: string): Observable<ApiResponseDTO> {
    return this.http.get<ApiResponseDTO>(`${this.baseUrl}/confirm-email`, {
      params: { userId, token },
    });
  }

  //POST /api/auth/login
  login(dto: AuthDTO.LoginDTO): Observable<AuthDTO.AuthResponseDTO> {
    return this.http
      .post<AuthDTO.AuthResponseDTO>(`${this.baseUrl}/login`, dto)
      .pipe(tap((res) => { this.saveTokens(res.token, res.refreshToken); this._scheduleWarning?.(); }));
  }

  //POST /api/auth/refresh
  refresh(dto: AuthDTO.RefreshTokenDTO): Observable<AuthDTO.AuthResponseDTO> {
    return this.http
      .post<AuthDTO.AuthResponseDTO>(`${this.baseUrl}/refresh`, dto)
      .pipe(tap((res) => { this.saveTokens(res.token, res.refreshToken); this._scheduleWarning?.(); }));
  }

  //POST /api/auth/logout  — requires Bearer token (attach via interceptor)
  logout(): Observable<ApiResponseDTO> {
    return this.http
      .post<ApiResponseDTO>(`${this.baseUrl}/logout`, {})
      .pipe(tap(() => this.clearTokens()));
  }

  // POST /api/auth/change-password
  changePassword(currentPassword: string, newPassword: string): Observable<ApiResponseDTO> {
    return this.http.post<ApiResponseDTO>(`${this.baseUrl}/change-password`, {
      currentPassword,
      newPassword,
    });
  }

  //POST /api/auth/forgot-password
  forgotPassword(email: string): Observable<ApiResponseDTO> {
    const dto: AuthDTO.ForgotPasswordDTO = { email };
    return this.http.post<ApiResponseDTO>(`${this.baseUrl}/forgot-password`, dto);
  }

  //POST /api/auth/reset-password
  resetPassword(dto: AuthDTO.ResetPasswordDTO): Observable<ApiResponseDTO> {
    return this.http.post<ApiResponseDTO>(`${this.baseUrl}/reset-password`, dto);
  }

  //POST /api/auth/resend-confirmation
  resendConfirmation(email: string): Observable<ApiResponseDTO> {
    const dto: AuthDTO.ForgotPasswordDTO = { email };
    return this.http.post<ApiResponseDTO>(`${this.baseUrl}/resend-confirmation`, dto);
  }

  isAdmin(): boolean {
    const token = this.getToken();
    if (!token) return false;
    try {
      const payload = JSON.parse(atob(token.split('.')[1]));
      const roles = payload['http://schemas.microsoft.com/ws/2008/06/identity/claims/role'];
      return Array.isArray(roles) ? roles.includes('Admin') : roles === 'Admin';
    } catch {
      return false;
    }
  }

  //Token helpers

  saveTokens(token: string, refreshToken: string): void {
    localStorage.setItem('token', token);
    localStorage.setItem('refreshToken', refreshToken);
  }

  getToken(): string | null {
    return localStorage.getItem('token');
  }

  getRefreshToken(): string | null {
    return localStorage.getItem('refreshToken');
  }

  clearTokens(): void {
    localStorage.removeItem('token');
    localStorage.removeItem('refreshToken');
  }

  isLoggedIn(): boolean {
    return !!this.getToken();
  }
}